src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java

```diff
@@ -215,61 +221,140 @@ private void turnOffReadiness(Throwable exception){
+            "reply-text=PRECONDITION_FAILED"
+    );
+
+    private void addShutdownListenerToChannel(Channel channel, Boolean withRecovery) {
+        channel.addShutdownListener(cause -> {
+            LOGGER.debug("Closing the channel: ", cause);
+            if (!cause.isHardError() && cause.getReference() instanceof Channel) {
+                Channel channelCause = (Channel) cause.getReference();
+                Method reason = cause.getReason();
+                if (reason instanceof AMQImpl.Channel.Close) {
+                    var castedReason = (AMQImpl.Channel.Close) reason;
+                    if (castedReason.getReplyCode() != 200) {
+                        StringBuilder errorBuilder = new StringBuilder("RabbitMQ soft error occurred: ");
+                        castedReason.appendArgumentDebugStringTo(errorBuilder);
+                        errorBuilder.append(" on channel ");
+                        errorBuilder.append(channelCause);
+                        String errorString = errorBuilder.toString();
+                        LOGGER.warn(errorString);
+
+                        var pinIdToChannelHolder = getChannelHolderByChannel(channel);
+                        if (pinIdToChannelHolder == null) return;
+
+                        if (withRecovery && recoverableErrors.stream().anyMatch(errorString::contains)) {
+                            var pinId = pinIdToChannelHolder.getKey();
+                            var holder = pinIdToChannelHolder.getValue();
+                            recoverSubscriptionsOfChannel(pinId, channel, holder);
+                        }
+                    }
+                }
             }
         });
+    }
+
+    private @Nullable Map.Entry<PinId, ChannelHolder> getChannelHolderByChannel(Channel channel) {
+        return channelsByPin
+                .entrySet()
+                .stream()
+                .filter(entry -> channel == entry.getValue().channel)
+                .findAny()
+                .orElse(null);
+    }
 
-        if (this.connection instanceof Recoverable) {
-            Recoverable recoverableConnection = (Recoverable) this.connection;
-            RecoveryListener recoveryListener = new RecoveryListener() {
-                @Override
-                public void handleRecovery(Recoverable recoverable) {
-                    LOGGER.debug("Count tries to recovery connection reset to 0");
-                    connectionRecoveryAttempts.set(0);
-                    metrics.getReadinessMonitor().enable();
-                    LOGGER.debug("Set RabbitMQ readiness to true");
+    private void recoverSubscriptionsOfChannel(@NotNull final PinId pinId, Channel channel, @NotNull final ChannelHolder holder) {
+        channelChecker.execute(() ->
+            holder.getCallbacksForRecovery(channel).ifPresent(subscriptionCallbacks -> {
+                LOGGER.info("Changing channel for holder with pin id: " + pinId);
+                channelsByPin.remove(pinId);
```

Comment:
Maybe we should close broken channel or do the similar action to be sure that resources is closed

Reply:
I believe the channel is already closed when this code is executed, because this code is called from shutdown listener.

Reply:
let's add the close logic just in case